From 9fc2c315b6b3633667f6b0d5e39c69afb1d3a00c Mon Sep 17 00:00:00 2001
From: Mathieu Bivert <mathieu.bivert@intel.com>
Date: Thu, 23 May 2013 17:05:54 +0200
Subject: [PORT FROM MAIN] crashtool: implement timestamp in events

BZ: 110538

Use timestamp in ct_event. Also correct a bug and reject events which do
not have a submitter_name or an ev_name (to be coherent with assert in ct_handlers)

Change-Id: I848fbaa26a340708787caa79cb562fd579a4178a
Signed-off-by: Mathieu Bivert <mathieu.bivert@intel.com>
Reviewed-on: http://android.intel.com:8080/109329
Reviewed-by: cactus <cactus@intel.com>
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: Zurmely, PierreX <pierrex.zurmely@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 libc/include/ctm.h             |    2 --
 libc/include/lct.h             |   26 +++++++++++++++-----------
 libc/kernel/common/linux/kct.h |    2 +-
 3 files changed, 16 insertions(+), 14 deletions(-)

diff --git a/libc/include/ctm.h b/libc/include/ctm.h
index 025ffb5..6bff158 100644
--- a/libc/include/ctm.h
+++ b/libc/include/ctm.h
@@ -63,8 +63,6 @@ extern void print_ct_event(struct ct_event *ev);
 extern int ct_stat_handler(struct ct_event *ev);
 extern int ct_crash_handler(struct ct_event *ev);
 
-extern struct ct_monitor_ctx ctm_ctx;
-
 extern ev_handler_t ev_handlers[CT_EV_LAST];
 
 #endif /* !CTM_H_ */
diff --git a/libc/include/lct.h b/libc/include/lct.h
index 190a79d..4ca08d4 100644
--- a/libc/include/lct.h
+++ b/libc/include/lct.h
@@ -4,6 +4,7 @@
 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>
+#include <time.h>
 #include <unistd.h>
 
 #include <sys/types.h>
@@ -69,23 +70,26 @@
 		} \
 	} while (0)
 
+/* keep ONE return to ensure inline */
 extern inline struct ct_event *lct_alloc_event(const char *submitter_name,
 					       const char *ev_name,
 					       enum ct_ev_type ev_type,
 					       unsigned int flags)
 {
-	struct ct_event *ev = calloc(1, sizeof(*ev));
-	if (ev) {
-		if (submitter_name)
-			strncpy(ev->submitter_name, submitter_name,
-				sizeof(ev->submitter_name));
-		if (ev_name)
-			strncpy(ev->ev_name, ev_name,
-				sizeof(ev->ev_name));
-		ev->timestamp = 0x42;
-		ev->flags = flags;
-		ev->type = ev_type;
+	struct ct_event *ev = NULL;
+
+	if (submitter_name && ev_name) {
+		ev = calloc(1, sizeof(*ev));
+		if (ev)  {
+			strncpy(ev->submitter_name, submitter_name, sizeof(ev->submitter_name)-1);
+			strncpy(ev->ev_name, ev_name, sizeof(ev->ev_name)-1);
+		
+			ev->timestamp = time(NULL);
+			ev->flags = flags;
+			ev->type = ev_type;
+		}
 	}
+
 	return ev;
 }
 
diff --git a/libc/kernel/common/linux/kct.h b/libc/kernel/common/linux/kct.h
index a9eb766..c172474 100644
--- a/libc/kernel/common/linux/kct.h
+++ b/libc/kernel/common/linux/kct.h
@@ -44,7 +44,7 @@ struct ct_event {
 	char submitter_name[MAX_SB_N];
 	char ev_name[MAX_EV_N];
 	enum ct_ev_type type;
-	__u64 timestamp;
+	time_t timestamp;
 	__u32 attchmt_size; /* sizeof(all_attachments inc. padding) */
 	__u32 flags;
 	struct ct_attchmt attachments[];
-- 
1.7.4.1

